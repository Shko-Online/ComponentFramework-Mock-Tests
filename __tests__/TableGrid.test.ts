/*
    Copyright (c) 2022 Betim Beja and Shko Online LLC
    Licensed under the MIT license.
*/

import * as sinon from "sinon";

import { ComponentFrameworkMockGenerator, DataSetMock } from "@shko.online/componentframework-mock";
import { TableGrid } from "@albanian-xrm/test-components/TableGrid";
import {
  IInputs,
  IOutputs,
} from "@albanian-xrm/test-components/TableGrid/generated/ManifestTypes";
import resource from "@albanian-xrm/test-components/TableGrid/strings/TableGrid.1033.resx";

describe("TableGrid", () => {
  let mockGenerator: ComponentFrameworkMockGenerator<IInputs, IOutputs>;
  const itemsLogicalName = "!!!items";
  const rows = [
    {
      myId: "1",
      alias: "First Item",
      alias2: "Second Item",
    },
    {
      myId: "2",
      alias: "First Item 2",
      alias2: "Second Item 2",
    },
    {
      myId: "3",
      alias: "First Item 3",
      alias2: "Second Item 3",
    },
  ];
  beforeEach(() => {
    const container = document.createElement("div");

    mockGenerator = new ComponentFrameworkMockGenerator(
      TableGrid,
      {
        simpleTableGrid: DataSetMock,
      },
      container
    );
    mockGenerator.SetControlResource(resource);
    const controlValue = mockGenerator.context.parameters
      .simpleTableGrid as DataSetMock;
    controlValue.columns = [
      {
        alias: "alias",
        dataType: "string",
        displayName: "Mocked Column",
        name: "alias",
        order: 1,
        visualSizeFactor: 200,
      },
      {
        alias: "alias2",
        dataType: "string",
        displayName: "Second Mocked Column",
        name: "alias2",
        order: 2,
        visualSizeFactor: 200,
      },
    ];
    mockGenerator.metadata.initMetadata([
      {
        EntitySetName: itemsLogicalName,
        LogicalName: itemsLogicalName,
        PrimaryIdAttribute: "myId",
        PrimaryNameAttribute: "alias",
        Attributes: ["myId", "alias", "alias2"].map(
          (logicalName) =>
            ({
              EntityLogicalName: itemsLogicalName,
              LogicalName: logicalName,
            } as ShkoOnline.StringAttributeMetadata)
        ),
      },
    ]);

    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.innerHTML = null;
  });
  it("Init should work", () => {
    mockGenerator.ExecuteInit();
    sinon.assert.calledOnce(mockGenerator.control.init);
    expect(document.body).toMatchSnapshot();
  });
  it("Update View should Work", () => {
    mockGenerator.ExecuteInit();
    mockGenerator.ExecuteUpdateView();
    sinon.assert.calledOnce(mockGenerator.control.init);
    sinon.assert.calledOnce(mockGenerator.control.updateView);
    expect(document.body).toMatchSnapshot();
  });
  it("GetOutputs should work", () => {
    mockGenerator.control.getOutputs();
    mockGenerator.ExecuteInit();
    mockGenerator.ExecuteUpdateView();
    expect(document.body).toMatchSnapshot();
  });
  it("Destroy should work", () => {
    mockGenerator.control.destroy();
    mockGenerator.ExecuteInit();
    mockGenerator.ExecuteUpdateView();
    expect(document.body).toMatchSnapshot();
  });
  it("Row Click Event handler for the associated row when being clicked", () => {
    mockGenerator.context._parameters.simpleTableGrid._Bind(
      itemsLogicalName,
      "items"
    );
    mockGenerator.context._parameters.simpleTableGrid._InitItems(rows || []);
    mockGenerator.ExecuteInit();
    mockGenerator.ExecuteUpdateView();

    const select = mockGenerator.container.querySelectorAll(
      `[rowRecId='${rows[0].myId}']`
    )[0];

    var evt = document.createEvent("Event");
    evt.initEvent("click", false, true);
    select.dispatchEvent(evt);
    expect(document.body).toMatchSnapshot();
  });
  it("Load More Button", () => {
    mockGenerator.ExecuteInit();
    mockGenerator.ExecuteUpdateView();
    const select = mockGenerator.container.querySelectorAll(
      ".LoadMoreButton_Style"
    )[0];
    var evt = document.createEvent("Event");
    evt.initEvent("click", false, true);
    select.dispatchEvent(evt);

    mockGenerator.ExecuteUpdateView();
    expect(document.body).toMatchSnapshot();
  });
  it("toggle load more button if it has next page", () => {
    mockGenerator.context.parameters.simpleTableGrid.paging.hasNextPage = true;
    mockGenerator.ExecuteInit();
    mockGenerator.ExecuteUpdateView();

    const select = mockGenerator.container.querySelector(
      ".SimpleTable_MainContainer_Style"
    );
    var evt = document.createEvent("Event");
    evt.initEvent("click", false, true);
    select.dispatchEvent(evt);
    expect(document.body).toMatchSnapshot();
  });
});
