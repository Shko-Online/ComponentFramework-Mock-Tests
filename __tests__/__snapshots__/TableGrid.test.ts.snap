// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableGrid Destroy should work 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
      style="height: 698px;"
    >
      <table
        class="SimpleTable_Table_Style"
      >
        <thead>
          <tr
            class="SimpleTable_TableRow_Style"
          >
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -125px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -126px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="No_Record_Style"
              colspan="2"
            />
          </tr>
        </tbody>
      </table>
      <button
        class="LoadMoreButton_Hidden_Style LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`TableGrid GetOutputs should work 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
      style="height: 698px;"
    >
      <table
        class="SimpleTable_Table_Style"
      >
        <thead>
          <tr
            class="SimpleTable_TableRow_Style"
          >
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -125px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -126px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="No_Record_Style"
              colspan="2"
            />
          </tr>
        </tbody>
      </table>
      <button
        class="LoadMoreButton_Hidden_Style LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`TableGrid Init should work 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
    >
      <table
        class="SimpleTable_Table_Style"
      />
      <button
        class="LoadMoreButton_Hidden_Style LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`TableGrid Load More Button 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
      style="height: 698px;"
    >
      <table
        class="SimpleTable_Table_Style"
      >
        <thead>
          <tr
            class="SimpleTable_TableRow_Style"
          >
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -125px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -126px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="No_Record_Style"
              colspan="2"
            />
          </tr>
        </tbody>
      </table>
      <button
        class="LoadMoreButton_Hidden_Style LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`TableGrid Row Click Event handler for the associated row when being clicked 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
      style="height: 698px;"
    >
      <table
        class="SimpleTable_Table_Style"
      >
        <thead>
          <tr
            class="SimpleTable_TableRow_Style"
          >
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -83px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="SimpleTable_TableRow_Style"
            rowrecid="1"
          >
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </td>
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </td>
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -83px;"
              />
            </td>
          </tr>
          <tr
            class="SimpleTable_TableRow_Style"
            rowrecid="2"
          >
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </td>
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </td>
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -83px;"
              />
            </td>
          </tr>
          <tr
            class="SimpleTable_TableRow_Style"
            rowrecid="3"
          >
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </td>
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -84px;"
              />
            </td>
            <td
              class="SimpleTable_TableCell_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -83px;"
              />
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="LoadMoreButton_Hidden_Style LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`TableGrid Update View should Work 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
      style="height: 698px;"
    >
      <table
        class="SimpleTable_Table_Style"
      >
        <thead>
          <tr
            class="SimpleTable_TableRow_Style"
          >
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -125px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -126px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="No_Record_Style"
              colspan="2"
            />
          </tr>
        </tbody>
      </table>
      <button
        class="LoadMoreButton_Hidden_Style LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`TableGrid toggle load more button if it has next page 1`] = `
<body>
  <div>
    <div
      class="SimpleTable_MainContainer_Style"
      style="height: 698px;"
    >
      <table
        class="SimpleTable_Table_Style"
      >
        <thead>
          <tr
            class="SimpleTable_TableRow_Style"
          >
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -125px;"
              />
            </th>
            <th
              class="SimpleTable_TableHeader_Style"
            >
              <div
                class="SimpleTable_TableCellInnerDiv_Style"
                style="max-width: -126px;"
              />
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="No_Record_Style"
              colspan="2"
            />
          </tr>
        </tbody>
      </table>
      <button
        class="LoadMoreButton_Style"
        type="button"
      />
    </div>
  </div>
</body>
`;
